grafana/grafana
# Azure Monitor: Deep linking from Log Analytic queries to the Azure Portal

# How to test

## Setup
1. Credentials can be found in the [private plugin-provisioning repo](https://github.com/grafana/plugin-provisioning/blob/master/plugins/azuremonitor/provisioning/datasources/azmonitor-ds.yaml)
2. Provision the YAML by placing it in the `conf/provisioning/datasources` and restarting your Grafana instance
3. Import this [sample dashboard](https://gist.github.com/daniellee/f6cfcc2b3afff6f2a2f9b31d353d803e) 

## Test

1. Click on a graph panel in the imported dashboard and click on the View in Azure Portal link
    ![image](https://user-images.githubusercontent.com/434655/81385323-7154a400-9113-11ea-9984-b28ec30e42af.png)
2. Verify that the link opens the query editor for Log Analytics in the Azure Portal with the query from the panel:

If you have never logged into the Azure Portal or use the Azure Log Analytics service before, you might not come directly to the query editor and instead see the below welcome screen. Click on `Get Started` and it should take you to the query editor:

![image](https://user-images.githubusercontent.com/434655/81385648-035cac80-9114-11ea-90c3-9b870be3c982.png)

## Notes to reviewer

- To build the url, 4 fields are needed - SubscriptionId, Resource Group, Workspace and the query. Unfortunately, the resource group is never used otherwise for Log Analytics and is therefore not saved in the panel json. I added another api call to fetch the workspace details and parse the resource group out of the id. I couldn't find a better way to do it - this is the only API call that I could find that returns anything to do with a workspace's resource group.
- I added some simple caching on the frontend for this extra call to get the workspace details. Is there a better way to do this? With the current solution, there is a cache per panel. If I have 5 panels on a dashboard, then the first time the dashboard loads, it will make 5 calls to get workspace details - after that they will be cached. 
    - I tried to do some caching with the `Cache-Control` HTTP header but couldn't get it to work reliably.
    - If I move the workspaces query and cache to the backend, then you lose the ability to refresh the page to invalidate the cache (for when you have added a new workspace).
- The query string is gzipped and base64 encoded on the backend but is url-encoded on the frontend. When I tried using the url-encoded version of base64 in Go, the Azure Portal couldn't unzip it.

Fixes #24321
